### PR TITLE
[WIP] Add git graph and trainer config

### LINKED ISSUE #191: Git-tree like for display model versioning
#### Problem description
As we have convert models from an original version, it could be nice and intuitive if we record this version history.
image

#### Steps to Reproduce the Problem
N.A. New feature.

#### Expected Behavior
Display the history tree of a bunch of models that all generated from one parent.

#### Other Information 
N.A.

### PR DESCRIPTION
**Pull request by univerone**
Closes #191

#### What has been done to verify that this works as intended?

- Add `parent_model_id` parameter to when registering models in modelhub 
- Add VersionTree page to track the origin of model

Here is a preview, snapshot will be added to frontend readme after fully connected to backend.

- Version Tree
Screenshot_2021-02-14 ML ModelCI(1)

#### Before submitting this PR, please make sure you have:

- [ ] run `python -m pytest tests/` and confirmed all checks still pass.
- [ ] verified that any code or assets from external sources are properly credited.

### CODE REVIEW

File: modelci/hub/manager.py [hunk 1]
```diff
@@ -405,3 +411,23 @@ def retrieve_model_by_task(task: Task) -> List[ModelBO]:
...
+
+    Returns:
+        List[ModelBO]: A list of model business object.
+    """
+    models = ModelService.get_models_by_parent_id(parent_id)
+    # check if found
+    if len(models) == 0:
+        raise FileNotFoundError('Model not found!')
+
+    _get_remote_model_weights(models)
```
Comment (YuanmingLeee) [inline modelci/hub/manager.py:431]: When you get the child models, do you really need to get them downloaded?

File: modelci/types/bo/model_bo.py [hunk 2]
```diff
@@ -56,6 +56,7 @@ def __init__(
             dataset: str,
             metric: Dict[Metric, float],
             task: Task,
+            parent_model_id: str,
```
Comment (YuanmingLeee) [inline modelci/types/bo/model_bo.py:59]: Note that this file should do removed in a few releases. We do not follow a Java design pattern anymore for more readability.

File: modelci/types/do/model_do.py [hunk 3]
```diff
@@ -46,6 +46,8 @@ class ModelDO(Document):
     weight = FileField()
     # Model task
     task = IntField(required=True)
+    # Parent Model ID
+    parent_model_id = StringField(required=True)
```
Comment (YuanmingLeee) [inline modelci/types/do/model_do.py:50]: This field is nullable. The 'root' model does not have a parent.

File: frontend/src/pages/GitTree/index.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
+import React from 'react';
```
Comment (YuanmingLeee) [inline frontend/src/pages/GitTree/index.tsx:1]: Rename the page to VersionTree

File: frontend/src/routes.ts [hunk 5]
```diff
@@ -31,6 +32,10 @@ const routerConfig = [
         path: '/visualizer/:id',
         component: Visualizer,
       },
+      {
+        path: '/gittree',
```
Comment (YuanmingLeee) [inline frontend/src/routes.ts:36]: path: '/version-tree'

File: modelci/types/vo/model_vo.py [hunk 6]
```diff
@@ -246,10 +250,25 @@ class ModelIn(BaseModel):
...
+    parent_model_id: str
     inputs: List[IOShapeVO]
     outputs: List[IOShapeVO]
     architecture: str
     framework: Framework
     version: int
     convert: bool
     profile: bool
+
+class TrainerConfig(BaseModel):
```
Comment (YuanmingLeee) [inline modelci/types/vo/model_vo.py:262]: Move this class to `trainer.py`. This class has nothing to do with the model object.

File: modelci/types/vo/model_vo.py [hunk 7]
```diff
@@ -215,6 +217,7 @@ class ModelDetailOut(BaseModel):
     dataset: str
     metric: Dict[Metric, float]
     task: Task
+    parent_model_id: str
```
Comment (YuanmingLeee) [inline modelci/types/vo/model_vo.py:220]: Optional field. The reason stated at `model_do.py`.

File: modelci/types/vo/model_vo.py [hunk 8]
```diff
@@ -179,6 +179,7 @@ class ModelListOut(BaseModel):
...
```
Comment (YuanmingLeee) [inline modelci/types/vo/model_vo.py:182]: Optional field.

File: modelci/types/vo/model_vo.py [hunk 9]
```diff
@@ -246,10 +250,25 @@ class ModelIn(BaseModel):
...
```
Comment (YuanmingLeee) [inline modelci/types/vo/model_vo.py:253]: Optional field.

File: frontend/src/pages/GitTree/index.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+      displayHash: false
+    },
+  },
+});
+
+
+const generateGitData = (model) => {
+  let data = {
+    author: {
+      name: "",
```
Comment (YuanmingLeee) [inline frontend/src/pages/GitTree/index.tsx:19]: Can use the model creator as the author

File: frontend/src/pages/GitTree/index.tsx [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+  public render() {
+    return (
+      <div>
+        <Gitgraph
+          options={{ template: gitGraphOption }}
+        >
+          {(gitgraph) => {
+            // this.generateTree(gitgraph)
+            var master = gitgraph.branch('PyTorch Model')
```
Comment (YuanmingLeee) [inline frontend/src/pages/GitTree/index.tsx:87]: Is this mocked?

File: frontend/src/pages/Visualizer/index.tsx [hunk 12]
```diff
@@ -1,43 +1,170 @@
...
 
-const onFinish = (values: any) => {
-	console.log('Success:', values);
-};
-
 const onFinishFailed = (errorInfo: any) => {
 	console.log('Failed:', errorInfo);
 };
 
+const defaultTrainerConfig = {
```
Comment (YuanmingLeee) [inline frontend/src/pages/Visualizer/index.tsx:25]: Why training logic is inside the visualizer page presenter?

Comment (YuanmingLeee) [reply frontend/src/pages/Visualizer/index.tsx:25]: Better put training parameter config and validation in a separate file.

**Review by YuanmingLeee — CHANGES_REQUESTED** [review]
In general, LGTM. There are some issues to be noted:
1. Model version graph and training config are two separate features, please submit two PRs next time if you are in this scenario.
2. When you do fast prototyping, such as the training stuff, you can put them into the `experimental` folder.

File: modelci/hub/manager.py (hunk 1 above)
Comment (univerone) [reply modelci/hub/manager.py:431]: No need and the model object should be returned

File: modelci/types/do/model_do.py (hunk 3 above)
Comment (univerone) [reply modelci/types/do/model_do.py:50]: The default value of  'root' model's parent_model_id will be an empty string, and we could extract root models by querying an empty parent_model_id

Comment (univerone) [reply modelci/types/do/model_do.py:50]: > The default value of 'root' model's parent_model_id will be an empty string, and we could extract root models by querying an empty parent_model_id

Setting this field as nullable is more convenient, I will fix that.

File: frontend/src/pages/GitTree/index.tsx (hunk 11 above)
Comment (univerone) [reply frontend/src/pages/GitTree/index.tsx:87]: Yes, this is mock data

**Pull request closed without merging**